sdk/dotnet/cmd/pulumi-language-dotnet/main.go

```diff
@@ -117,37 +118,231 @@ func newLanguageHost(exec, engineAddress, tracing string) pulumirpc.LanguageRunt
 }
 
 // GetRequiredPlugins computes the complete set of anticipated plugins required by a program.
-func (host *dotnetLanguageHost) GetRequiredPlugins(ctx context.Context,
+func (host *dotnetLanguageHost) GetRequiredPlugins(
+	ctx context.Context,
 	req *pulumirpc.GetRequiredPluginsRequest) (*pulumirpc.GetRequiredPluginsResponse, error) {
-	// TODO: implement this.
-	return &pulumirpc.GetRequiredPluginsResponse{}, nil
-}
 
-// RPC endpoint for LanguageRuntimeServer::Run
-func (host *dotnetLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest) (*pulumirpc.RunResponse, error) {
+	logging.V(5).Infof("GetRequiredPlugins: %v", req.GetProgram())
+
+	// First do a `dotnet build`.  This will ensure that all the nuget dependencies of the project
+	// are restored and locally available for us.
 	if err := host.DotnetBuild(ctx, req); err != nil {
-		return &pulumirpc.RunResponse{Error: err.Error()}, nil
+		return nil, err
+	}
+
+	// now, introspect the user project to see which pulumi resource packages it references.
+	pulumiPackages, err := host.DeterminePulumiPackages(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure we know where the local nuget package cache directory is.  User can specify where that
+	// is located, so this makes sure we respect any custom location they may have.
+	packageDir, err := host.DetermineDotnetPackageDirectory(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Now that we know the set of pulumi packages referenced and we know where packages have been restored to,
+	// we can examine each package to determine the corresponding resource-plugin for it.
+
+	plugins := []*pulumirpc.PluginDependency{}
+	packageToVersion := make(map[string]string)
+	for _, parts := range pulumiPackages {
+		packageName := parts[0]
+		packageVersion := parts[1]
+
+		if existingVersion := packageToVersion[packageName]; existingVersion == packageVersion {
+			// only include distinct dependencies.
```

Comment:
Do we need to support SxS scenarios?

Reply:
it is *waaaay* too early to ask that now.  i'm at least making sure that if there is a ref that says "i need htis plugin" that we require the plugin.  I'm sure we'll have all sorts of painful sxs stuff we'll have to work through.

Reply:
It's definitely not too early to ask that. We don't have to support them now, but we absolutely need to start thinking through whether or not we will. And this is a land mine if we do need to in the fullness of time.

Reply:
let me rephrase.  I think the topic is important (and i'll add to the "future work" list).  I just don't think it's something to figure out in this PR :)